synapse/storage/devices.py

```diff
@@ -582,11 +603,11 @@ def delete_device(self, user_id, device_id):
         Returns:
             defer.Deferred
         """
-        yield self._simple_delete_one(
-            table="devices",
-            keyvalues={"user_id": user_id, "device_id": device_id},
-            desc="delete_device",
-        )
+        sql = """
+            DELETE FROM devices
+            WHERE user_id = ? AND device_id = ? AND NOT COALESCE(hidden, ?)
```

Comment:
having looked at how this works out, I'm really sorry but I think I'm going to revise my original comments here; it's just too awkward. I think the 30s or so it's going to take to add a NOT NULL column to the table will be worth it for not jumping through these hoops and having to manage a tristate.

Really sorry for vacillating on this.